File: backend/app/api/v1/reviews.py
"""Review API endpoints"""

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional
import uuid

from app.api.v1.deps import get_db, get_current_user
from app.models.user import User
from app.models.review import ReviewType
from app.schemas.review import (
    ReviewCreate,
    ReviewUpdate,
    ReviewResponse,
    ReviewListResponse,
    MarkReviewHelpfulResponse
)
from app.schemas.common import APIResponse
from app.services.review_service import review_service


router = APIRouter()


@router.post("", response_model=APIResponse[ReviewResponse])
async def create_review(
    review_data: ReviewCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Create a new review
    
    Reviews can be for:
    - Billboards (requires billboard_id)
    - Owners (requires reviewee_id)
    - Advertisers (requires reviewee_id)
    """
    review = await review_service.create_review(
        db=db,
        user_id=current_user.id,
        review_data=review_data
    )
    
    return APIResponse(
        success=True,
        data=review,
        message="Review created successfully"
    )


@router.put("/{review_id}", response_model=APIResponse[ReviewResponse])
async def update_review(
    review_id: str,
    review_data: ReviewUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Update an existing review
    
    Only the review creator can update their review.
    """
    review = await review_service.update_review(
        db=db,
        review_id=uuid.UUID(review_id),
        user_id=current_user.id,
        review_data=review_data
    )
    
    return APIResponse(
        success=True,
        data=review,
        message="Review updated successfully"
    )


@router.delete("/{review_id}", response_model=APIResponse[dict])
async def delete_review(
    review_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Delete a review
    
    Only the review creator can delete their review.
    """
    await review_service.delete_review(
        db=db,
        review_id=uuid.UUID(review_id),
        user_id=current_user.id
    )
    
    return APIResponse(
        success=True,
        data={"deleted": True},
        message="Review deleted successfully"
    )


@router.get("", response_model=APIResponse[ReviewListResponse])
async def get_reviews(
    review_type: Optional[ReviewType] = Query(None, description="Filter by review type"),
    billboard_id: Optional[str] = Query(None, description="Filter by billboard ID"),
    reviewee_id: Optional[str] = Query(None, description="Filter by reviewee (user) ID"),
    min_rating: Optional[int] = Query(None, ge=1, le=5, description="Minimum rating filter"),
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Items per page"),
    db: AsyncSession = Depends(get_db)
):
    """
    Get reviews with optional filters
    
    Filters:
    - review_type: billboard, owner, or advertiser
    - billboard_id: Get reviews for a specific billboard
    - reviewee_id: Get reviews for a specific user (owner or advertiser)
    - min_rating: Minimum star rating (1-5)
    """
    reviews = await review_service.get_reviews(
        db=db,
        review_type=review_type,
        billboard_id=uuid.UUID(billboard_id) if billboard_id else None,
        reviewee_id=uuid.UUID(reviewee_id) if reviewee_id else None,
        min_rating=min_rating,
        page=page,
        page_size=page_size
    )
    
    return APIResponse(
        success=True,
        data=reviews,
        message="Reviews retrieved successfully"
    )


@router.post("/{review_id}/helpful", response_model=APIResponse[MarkReviewHelpfulResponse])
async def mark_review_helpful(
    review_id: str,
    db: AsyncSession = Depends(get_db)
):
    """
    Mark a review as helpful
    
    Increments the helpful count for the review.
    No authentication required to mark reviews as helpful.
    """
    helpful_count = await review_service.mark_review_helpful(
        db=db,
        review_id=uuid.UUID(review_id)
    )
    
    return APIResponse(
        success=True,
        data=MarkReviewHelpfulResponse(
            success=True,
            helpful_count=helpful_count
        ),
        message="Review marked as helpful"
    )


@router.get("/billboard/{billboard_id}/summary", response_model=APIResponse[dict])
async def get_billboard_review_summary(
    billboard_id: str,
    db: AsyncSession = Depends(get_db)
):
    """
    Get review summary for a billboard
    
    Returns:
    - Average rating
    - Total review count
    - Rating distribution (1-5 stars)
    """
    reviews = await review_service.get_reviews(
        db=db,
        review_type=ReviewType.BILLBOARD,
        billboard_id=uuid.UUID(billboard_id),
        page=1,
        page_size=1000  # Get all reviews for summary
    )
    
    # Calculate rating distribution
    rating_distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
    for review in reviews.reviews:
        rating_distribution[review.rating] += 1
    
    summary = {
        "average_rating": reviews.average_rating,
        "total_reviews": reviews.total,
        "rating_distribution": rating_distribution
    }
    
    return APIResponse(
        success=True,
        data=summary,
        message="Billboard review summary retrieved"
    )


@router.get("/user/{user_id}/summary", response_model=APIResponse[dict])
async def get_user_review_summary(
    user_id: str,
    db: AsyncSession = Depends(get_db)
):
    """
    Get review summary for a user (owner or advertiser)
    
    Returns:
    - Average rating
    - Total review count
    - Rating distribution (1-5 stars)
    """
    reviews = await review_service.get_reviews(
        db=db,
        reviewee_id=uuid.UUID(user_id),
        page=1,
        page_size=1000  # Get all reviews for summary
    )
    
    # Calculate rating distribution
    rating_distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
    for review in reviews.reviews:
        rating_distribution[review.rating] += 1
    
    summary = {
        "average_rating": reviews.average_rating,
        "total_reviews": reviews.total,
        "rating_distribution": rating_distribution
    }
    
    return APIResponse(
        success=True,
        data=summary,
        message="User review summary retrieved"
    )

File: client/app/api/endpoints/reviews.ts
import apiClient from '../index'
import type { 
  Review,
  ReviewCreateInput,
  ReviewUpdateInput,
  ReviewSummary,
} from '@/lib/types/review'
import type { ApiResponse, PaginatedResponse } from '@/lib/types'

// Helper: Convert camelCase to snake_case
const toSnakeCase = (obj: Record<string, any>): Record<string, any> => {
  const result: Record<string, any> = {}
  for (const [key, value] of Object.entries(obj)) {
    if (value === undefined) continue
    
    const snakeKey = key.replace(/[A-Z]/g, letter => `_${letter.toLowerCase()}`)
    result[snakeKey] = value
  }
  return result
}

// Helper: Convert snake_case to camelCase
const toCamelCase = (obj: Record<string, any>): Record<string, any> => {
  const result: Record<string, any> = {}
  for (const [key, value] of Object.entries(obj)) {
    if (value === null || value === undefined) {
      result[key] = value
      continue
    }
    
    const camelKey = key.replace(/_([a-z])/g, (_, letter) => letter.toUpperCase())
    
    if (typeof value === 'object' && !Array.isArray(value) && !(value instanceof Date)) {
      result[camelKey] = toCamelCase(value)
    } else if (Array.isArray(value)) {
      result[camelKey] = value.map(item => 
        typeof item === 'object' && item !== null ? toCamelCase(item) : item
      )
    } else {
      result[camelKey] = value
    }
  }
  return result
}

// Transform review from API
const transformReview = (data: any): Review => {
  return toCamelCase(data) as Review
}

/**
 * Create a new review
 */
export const createReview = async (data: ReviewCreateInput): Promise<ApiResponse<Review>> => {
  const snakeData = toSnakeCase(data)
  const response = await apiClient.post<ApiResponse<any>>('/reviews', snakeData)
  
  return {
    ...response.data,
    data: transformReview(response.data.data),
  }
}

/**
 * Update an existing review
 */
export const updateReview = async (
  reviewId: string,
  data: ReviewUpdateInput
): Promise<ApiResponse<Review>> => {
  const snakeData = toSnakeCase(data)
  const response = await apiClient.patch<ApiResponse<any>>(`/reviews/${reviewId}`, snakeData)
  
  return {
    ...response.data,
    data: transformReview(response.data.data),
  }
}

/**
 * Delete a review
 */
export const deleteReview = async (reviewId: string): Promise<ApiResponse<{ message: string }>> => {
  const response = await apiClient.delete<ApiResponse<{ message: string }>>(`/reviews/${reviewId}`)
  return response.data
}

/**
 * Get reviews for a billboard
 */
export const getBillboardReviews = async (
  billboardId: string,
  page: number = 1,
  limit: number = 20
): Promise<ApiResponse<PaginatedResponse<Review>>> => {
  const response = await apiClient.get<ApiResponse<any>>(`/reviews/billboard/${billboardId}`, {
    params: { page, limit },
  })
  
  return {
    ...response.data,
    data: {
      items: response.data.data.items.map(transformReview),
      total: response.data.data.total,
      page: response.data.data.page,
      limit: response.data.data.limit,
      pages: response.data.data.total_pages,
    },
  }
}

/**
 * Get reviews for a user (owner or advertiser)
 */
export const getUserReviews = async (
  userId: string,
  page: number = 1,
  limit: number = 20
): Promise<ApiResponse<PaginatedResponse<Review>>> => {
  const response = await apiClient.get<ApiResponse<any>>(`/reviews/user/${userId}`, {
    params: { page, limit },
  })
  
  return {
    ...response.data,
    data: {
      items: response.data.data.items.map(transformReview),
      total: response.data.data.total,
      page: response.data.data.page,
      limit: response.data.data.limit,
      pages: response.data.data.total_pages,
    },
  }
}

/**
 * Get current user's reviews (reviews they've written)
 */
export const getMyReviews = async (
  page: number = 1,
  limit: number = 20
): Promise<ApiResponse<PaginatedResponse<Review>>> => {
  const response = await apiClient.get<ApiResponse<any>>('/reviews/me', {
    params: { page, limit },
  })
  
  return {
    ...response.data,
    data: {
      items: response.data.data.items.map(transformReview),
      total: response.data.data.total,
      page: response.data.data.page,
      limit: response.data.data.limit,
      pages: response.data.data.total_pages,
    },
  }
}

/**
 * Get review summary for a billboard
 */
export const getBillboardReviewSummary = async (
  billboardId: string
): Promise<ApiResponse<ReviewSummary>> => {
  const response = await apiClient.get<ApiResponse<any>>(`/reviews/billboard/${billboardId}/summary`)
  
  return {
    ...response.data,
    data: {
      averageRating: response.data.data.average_rating,
      totalReviews: response.data.data.total_reviews,
      ratingDistribution: response.data.data.rating_distribution,
    },
  }
}

/**
 * Get review summary for a user (owner or advertiser)
 */
export const getUserReviewSummary = async (
  userId: string
): Promise<ApiResponse<ReviewSummary>> => {
  const response = await apiClient.get<ApiResponse<any>>(`/reviews/user/${userId}/summary`)
  
  return {
    ...response.data,
    data: {
      averageRating: response.data.data.average_rating,
      totalReviews: response.data.data.total_reviews,
      ratingDistribution: response.data.data.rating_distribution,
    },
  }
}

/**
 * Mark a review as helpful
 */
export const markReviewHelpful = async (reviewId: string): Promise<ApiResponse<Review>> => {
  const response = await apiClient.post<ApiResponse<any>>(`/reviews/${reviewId}/helpful`)
  
  return {
    ...response.data,
    data: transformReview(response.data.data),
  }
}

/**
 * Unmark a review as helpful
 */
export const unmarkReviewHelpful = async (reviewId: string): Promise<ApiResponse<Review>> => {
  const response = await apiClient.delete<ApiResponse<any>>(`/reviews/${reviewId}/helpful`)
  
  return {
    ...response.data,
    data: transformReview(response.data.data),
  }
}

/**
 * Get all reviews with optional filters
 */
export const getAllReviews = async (
  page: number = 1,
  limit: number = 20,
  filters?: {
    rating?: number
    reviewType?: 'billboard' | 'owner' | 'advertiser'
    search?: string
  }
): Promise<ApiResponse<PaginatedResponse<Review>>> => {
  const params: any = { page, limit }
  
  if (filters?.rating) params.rating = filters.rating
  if (filters?.reviewType) params.review_type = filters.reviewType
  if (filters?.search) params.search = filters.search
  
  const response = await apiClient.get<ApiResponse<any>>('/reviews', { params })
  
  return {
    ...response.data,
    data: {
      items: response.data.data.items.map(transformReview),
      total: response.data.data.total,
      page: response.data.data.page,
      limit: response.data.data.limit,
      pages: response.data.data.total_pages,
    },
  }
}

/**
 * Get overall review summary (all reviews)
 */
export const getAllReviewsSummary = async (): Promise<ApiResponse<ReviewSummary>> => {
  const response = await apiClient.get<ApiResponse<any>>('/reviews/summary')
  
  return {
    ...response.data,
    data: {
      averageRating: response.data.data.average_rating,
      totalReviews: response.data.data.total_reviews,
      ratingDistribution: response.data.data.rating_distribution,
    },
  }
}

// Export all endpoints
export const reviewEndpoints = {
  createReview,
  updateReview,
  deleteReview,
  getBillboardReviews,
  getUserReviews,
  getMyReviews,
  getAllReviews,
  getBillboardReviewSummary,
  getUserReviewSummary,
  getAllReviewsSummary,
  markReviewHelpful,
  unmarkReviewHelpful,
}
